fix: Fixed media content upload to bucket

frgfm opened the pull request:
As witnessed over the past day, we've had some troubles with media upload. Some were fixed in #95 and #96 fortunately, but two issues remained: 
- adding the subfolder in the bucket name was not supported by the CSP client
- empty file could be uploaded, if the byte position of a file were to be offset for instance

This PR fixes both issues by adding a generalized bucket key resolution and resetting the byte position after hashing to avoid uploading empty files. I tried it on my end by deploying the feature branch on Heroku, and it works:
![qarnot_bucket](https://user-images.githubusercontent.com/26927750/101555245-8a2e7c00-39b8-11eb-9333-805d6e5f0ccf.png)


Feel free to try by yourself before the next deployment :ok_hand: 

Any feedback is welcome!

florianriche commented on src/app/api/routes/media.py:
```diff
@@ -94,7 +94,11 @@ async def upload_media(media_id: int = Path(..., gt=0),
+    bucket_key = prepend_bucket_folder(file_name)
 
     upload_success = await bucket_service.upload_file(bucket_key=bucket_key,
                                                       file_binary=file.file)
```
Au passage, can you also fix the return message below?
            detail="The upload did not succeed"
==>     detail="The download did not succeed"

frgfm replied:
Not sure I follow: here we're actually trying to upload. Did you mean in `get_media_image`?

florianriche replied:
Yes,(for some reason I can't comment directly to point this line in the Github interface)

frgfm replied:
Done!

florianriche reviewed (approved):
Looks good to me. Just a small additionnal fix to include in this PR while we are at it.

florianriche reviewed (dismissed):
Looks good! 
It's a bit unsettling because when I tried it on my side, I didn't see a new media appearing in Qarnot but that's because I was uploading the same image. But it worked!

florianriche commented on .github/workflows/main.yml:
```diff
@@ -78,7 +78,9 @@ jobs:
-        run: PORT=8002 docker-compose up -d --build
+        run: |
+          PORT=8002 docker-compose up -d --build
+          docker ps
```
docker ps is required?

frgfm replied:
cf. my comment :)

florianriche commented on .github/workflows/main.yml:
```diff
@@ -89,4 +91,6 @@ jobs:
       - name: End-to-End test
-        run: python scripts/api_e2e.py 8002
+        run: |
+          sleep 5
```
Why the sleep?

frgfm replied:
cf. my comment :)

florianriche commented on .github/workflows/client-main.yml:
```diff
@@ -55,18 +55,19 @@ jobs:
       - name: Run client unittests
         run: |
           coverage --version
+          sleep 5
```
why the sleep?

frgfm replied:
same

florianriche commented on .github/workflows/client-main.yml:
```diff
@@ -55,18 +55,19 @@ jobs:
+      - name: Build & run API server
+        run: |
+          PORT=8002 docker-compose up -d --build
+          docker ps
```
why the docker ps ?

frgfm replied:
same

frgfm commented:
Added a unittest for bucket key resolution (after all, I introduced a feature, this should be tested)
And additionally, I think I found the problem with `ConnectionError`: the tests were run right after the docker started, while the FastAPI server needs some time to run and become reachable :sweat_smile: 
I added a workflow fix (moving the docker start as early as possible, and add a sleep before running the tests) and it seems to be doing the trick :raised_hands:
